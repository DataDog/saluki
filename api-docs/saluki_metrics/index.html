<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Metrics-related helpers and utilities."><title>saluki_metrics - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="saluki_metrics" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (4b94758d2 2025-10-13)" data-channel="nightly" data-search-js="search-8d3311b9.js" data-stringdex-js="stringdex-828709d0.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Crate saluki_metrics</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../saluki_metrics/index.html">saluki_<wbr>metrics</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#macros">Crate Items</a></h3><ul class="block"><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>saluki_<wbr>metrics</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/saluki_metrics/lib.rs.html#1-37">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Metrics-related helpers and utilities.</p>
</div></details><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="macro" href="macro.static_metrics.html" title="macro saluki_metrics::static_metrics">static_<wbr>metrics</a></dt><dd>Creates statically-defined metrics within a dedicated container struct.</dd></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.MetricsBuilder.html" title="struct saluki_metrics::MetricsBuilder">Metrics<wbr>Builder</a></dt><dd>Builder for constructing metrics.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.MetricTag.html" title="trait saluki_metrics::MetricTag">Metric<wbr>Tag</a></dt><dd>A metric tag.</dd><dt><a class="trait" href="trait.Stringable.html" title="trait saluki_metrics::Stringable">Stringable</a></dt><dd>A type that can be converted into a <code>SharedString</code>.</dd></dl></section></div></main></body></html>