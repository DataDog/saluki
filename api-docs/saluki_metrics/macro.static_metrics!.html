<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.static_metrics.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.static_metrics.html">macro.static_metrics.html</a>...</p>
    <script>location.replace("macro.static_metrics.html" + location.search + location.hash);</script>
</body>
</html>