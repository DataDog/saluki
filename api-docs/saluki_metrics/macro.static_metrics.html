<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Creates statically-defined metrics within a dedicated container struct."><title>static_metrics in saluki_metrics - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="saluki_metrics" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (4068bafed 2025-10-20)" data-channel="nightly" data-search-js="search-8d3311b9.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">static_metrics</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../saluki_metrics/index.html">saluki_<wbr>metrics</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">static_<wbr>metrics</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#labels" title="Labels">Labels</a></li><li><a href="#levels" title="Levels">Levels</a></li><li><a href="#examples" title="Examples">Examples</a><ul><li><a href="#basic" title="Basic">Basic</a></li><li><a href="#customized-levels" title="Customized Levels">Customized Levels</a></li></ul></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate saluki_<wbr>metrics</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="index.html">saluki_metrics</a></div><h1>Macro <span class="macro">static_<wbr>metrics</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/saluki_metrics/macros.rs.html#163-221">Source</a> </span></div><pre class="rust item-decl"><code>macro_rules! static_metrics {
    (name =&gt; $name:ident, prefix =&gt; $prefix:ident, metrics =&gt; [$($metric_type:ident($metric_name:ident)),+ $(,)?] $(,)?) =&gt; { ... };
    (name =&gt; $name:ident, prefix =&gt; $prefix:ident, labels =&gt; [$($label_key:ident: $label_ty:ty),*], metrics =&gt; [$($metric_type:ident($metric_name:ident)),+ $(,)?] $(,)?) =&gt; { ... };
}</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Creates statically-defined metrics within a dedicated container struct.</p>
<p>In some cases, the metrics needed for a component are well-established and do not require a high-level of dynamism:
perhaps only a single label is needed for all metrics, and the value is known ahead of time. In these cases, it can
be useful to declare the metrics up front, and in a contained way, to avoid having to deal with the string-y metric
names (and any labels) at each and every callsite.</p>
<p><code>static_metrics!</code> allows defining a number of metrics contained within a single struct. The struct can be
initialized with a fixed set of labels, which is used when registering the metrics. The metrics can then be accessed
with simple accessors on the struct, allowing for ergonomic access from the calling code.</p>
<h2 id="labels"><a class="doc-anchor" href="#labels">§</a>Labels</h2>
<p>A fixed set of labels can be configured for all metrics that are registered. These labels have their definition
defined when calling <code>static_metrics!</code>, and the label value is provided when initializing the generated struct.</p>
<p>This allows for quickly applying the same set of labels to all metrics defined within the container struct, and
being able to handle them in a strong-typed way right up until the moment where they need to be rendered as strings.</p>
<h2 id="levels"><a class="doc-anchor" href="#levels">§</a>Levels</h2>
<p>In <code>metrics</code>, metrics have an inherent “level”, similar to logs: trace, debug, info, warn, and error.  Levels can be
used to filter metrics based on their importance or severity. For example, a trace-level metric might be only be
required for debugging and should not be sent all the time, while warn- or error-level metrics should be sent all the
time.</p>
<p>We expose the ability to specify the level to use for a metric by specifying it as a prefix to the metric type. See the
below examples for more details. Metrics can be defined at the trace, debug, or info level, and will default to the info
level if no level is specified.</p>
<h2 id="examples"><a class="doc-anchor" href="#examples">§</a>Examples</h2><h3 id="basic"><a class="doc-anchor" href="#basic">§</a>Basic</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="comment">// We are required to provide a name for the struct, as well as the metric prefix to apply to each of the defined metrics.
//
// Naturally, we also have to define metrics, but labels are optionally and can be excluded from the macro usage entirely.
</span><span class="macro">static_metrics!</span>(
   name =&gt; FrobulatorMetrics,
   prefix =&gt; frobulator,
   labels =&gt; [process_id: u32],
   metrics =&gt; [
       counter(successful_frobulations),
   ],
);

<span class="kw">struct </span>Frobulator {
    metrics: FrobulatorMetrics,
}

<span class="kw">impl </span>Frobulator {
    <span class="kw">fn </span>new(process_id: u32) -&gt; <span class="self">Self </span>{
        <span class="self">Self </span>{
           metrics: FrobulatorMetrics::new(process_id),
        }
    }

    <span class="kw">fn </span>frobulate(<span class="kw-2">&amp;</span><span class="self">self</span>) {
        <span class="comment">/* Do the frobulation...*/
        </span><span class="self">self</span>.metrics.successful_frobulations().increment(<span class="number">1</span>)
    }
}</code></pre></div><h3 id="customized-levels"><a class="doc-anchor" href="#customized-levels">§</a>Customized Levels</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code>
<span class="comment">// In this example, we define the level of metrics by prefixing them, such that we have two debug metrics
// (`tasks_preempted` and `pending_io_wakeups`) and one trace metric (`task_poll_duration`). Our non-prefixed
// metric, `tasks_completed`, defaults to the INFO level.
</span><span class="macro">static_metrics!</span>(
   name =&gt; RuntimeMetrics,
   prefix =&gt; runtime,
   metrics =&gt; [
       counter(tasks_completed),
       debug_counter(tasks_preempted),
       debug_gauge(pending_io_wakeups),
       trace_histogram(task_poll_duration),
   ],
);</code></pre></div></div></details></section></div></main></body></html>