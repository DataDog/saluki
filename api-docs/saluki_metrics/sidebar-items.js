window.SIDEBAR_ITEMS = {"macro":["static_metrics"],"struct":["MetricsBuilder"],"trait":["MetricTag","Stringable"]};