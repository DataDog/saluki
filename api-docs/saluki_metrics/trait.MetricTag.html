<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="A metric tag."><title>MetricTag in saluki_metrics - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="saluki_metrics" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (4068bafed 2025-10-20)" data-channel="nightly" data-search-js="search-8d3311b9.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">MetricTag</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../saluki_metrics/index.html">saluki_<wbr>metrics</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Metric<wbr>Tag</a></h2><h3><a href="#required-methods">Required Methods</a></h3><ul class="block"><li><a href="#tymethod.into_label" title="into_label">into_label</a></li></ul><h3><a href="#foreign-impls">Implementations on Foreign Types</a></h3><ul class="block"><li><a href="#impl-MetricTag-for-%26str" title="&#38;&#39;static str">&#38;&#39;static str</a></li><li><a href="#impl-MetricTag-for-(%26str,+T)" title="(&#38;&#39;static str, T)">(&#38;&#39;static str, T)</a></li><li><a href="#impl-MetricTag-for-String" title="String">String</a></li></ul><h3><a href="#implementors">Implementors</a></h3></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate saluki_<wbr>metrics</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="index.html">saluki_metrics</a></div><h1>Trait <span class="trait">Metric<wbr>Tag</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/saluki_metrics/builder.rs.html#21-29">Source</a> </span></div><pre class="rust item-decl"><code>pub trait MetricTag: Sealed {
    // Required method
    fn <a href="#tymethod.into_label" class="fn">into_label</a>(self) -&gt; <a class="struct" href="https://docs.rs/metrics/0.24.2/metrics/label/struct.Label.html" title="struct metrics::label::Label">Label</a>;
}</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>A metric tag.</p>
<p>Marker trait for types which can support being used as a metric tag, in order to optimize their internal
representation to avoid unnecessary allocations.</p>
<p>This trait is sealed and cannot be implemented outside of this crate.</p>
</div></details><h2 id="required-methods" class="section-header">Required Methods<a href="#required-methods" class="anchor">§</a></h2><div class="methods"><details class="toggle method-toggle" open><summary><section id="tymethod.into_label" class="method"><a class="src rightside" href="../src/saluki_metrics/builder.rs.html#28">Source</a><h4 class="code-header">fn <a href="#tymethod.into_label" class="fn">into_label</a>(self) -&gt; <a class="struct" href="https://docs.rs/metrics/0.24.2/metrics/label/struct.Label.html" title="struct metrics::label::Label">Label</a></h4></section></summary><div class="docblock"><p>Consumes <code>self</code> and converts it to a tag.</p>
<p>Under the hood, the <a href="https://docs.rs/metrics"><code>metrics</code></a> crate is used, which calls tags “labels” instead, which is where the
return type naming comes from.</p>
</div></details></div><h2 id="foreign-impls" class="section-header">Implementations on Foreign Types<a href="#foreign-impls" class="anchor">§</a></h2><details class="toggle implementors-toggle"><summary><section id="impl-MetricTag-for-%26str" class="impl"><a class="src rightside" href="../src/saluki_metrics/builder.rs.html#31-38">Source</a><a href="#impl-MetricTag-for-%26str" class="anchor">§</a><h3 class="code-header">impl <a class="trait" href="trait.MetricTag.html" title="trait saluki_metrics::MetricTag">MetricTag</a> for &amp;'static <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.str.html">str</a></h3></section></summary><div class="impl-items"><section id="method.into_label" class="method trait-impl"><a class="src rightside" href="../src/saluki_metrics/builder.rs.html#32-37">Source</a><a href="#method.into_label" class="anchor">§</a><h4 class="code-header">fn <a href="#tymethod.into_label" class="fn">into_label</a>(self) -&gt; <a class="struct" href="https://docs.rs/metrics/0.24.2/metrics/label/struct.Label.html" title="struct metrics::label::Label">Label</a></h4></section></div></details><details class="toggle implementors-toggle"><summary><section id="impl-MetricTag-for-String" class="impl"><a class="src rightside" href="../src/saluki_metrics/builder.rs.html#40-47">Source</a><a href="#impl-MetricTag-for-String" class="anchor">§</a><h3 class="code-header">impl <a class="trait" href="trait.MetricTag.html" title="trait saluki_metrics::MetricTag">MetricTag</a> for <a class="struct" href="https://doc.rust-lang.org/nightly/alloc/string/struct.String.html" title="struct alloc::string::String">String</a></h3></section></summary><div class="impl-items"><section id="method.into_label-1" class="method trait-impl"><a class="src rightside" href="../src/saluki_metrics/builder.rs.html#41-46">Source</a><a href="#method.into_label-1" class="anchor">§</a><h4 class="code-header">fn <a href="#tymethod.into_label" class="fn">into_label</a>(self) -&gt; <a class="struct" href="https://docs.rs/metrics/0.24.2/metrics/label/struct.Label.html" title="struct metrics::label::Label">Label</a></h4></section></div></details><details class="toggle implementors-toggle"><summary><section id="impl-MetricTag-for-(%26str,+T)" class="impl"><a class="src rightside" href="../src/saluki_metrics/builder.rs.html#49-56">Source</a><a href="#impl-MetricTag-for-(%26str,+T)" class="anchor">§</a><h3 class="code-header">impl&lt;T&gt; <a class="trait" href="trait.MetricTag.html" title="trait saluki_metrics::MetricTag">MetricTag</a> for (&amp;'static <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.str.html">str</a>, T)<div class="where">where
    T: <a class="trait" href="https://doc.rust-lang.org/nightly/core/convert/trait.Into.html" title="trait core::convert::Into">Into</a>&lt;<a class="type" href="https://docs.rs/metrics/0.24.2/metrics/common/type.SharedString.html" title="type metrics::common::SharedString">SharedString</a>&gt;,</div></h3></section></summary><div class="impl-items"><section id="method.into_label-2" class="method trait-impl"><a class="src rightside" href="../src/saluki_metrics/builder.rs.html#53-55">Source</a><a href="#method.into_label-2" class="anchor">§</a><h4 class="code-header">fn <a href="#tymethod.into_label" class="fn">into_label</a>(self) -&gt; <a class="struct" href="https://docs.rs/metrics/0.24.2/metrics/label/struct.Label.html" title="struct metrics::label::Label">Label</a></h4></section></div></details><h2 id="implementors" class="section-header">Implementors<a href="#implementors" class="anchor">§</a></h2><div id="implementors-list"></div><script src="../trait.impl/saluki_metrics/builder/trait.MetricTag.js" data-ignore-extern-crates="std,alloc" async></script></section></div></main></body></html>