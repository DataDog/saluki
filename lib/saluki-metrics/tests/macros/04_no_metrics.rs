use saluki_metrics::static_metrics;

static_metrics! {
	name => Metrics,
	prefix => basic_usage,
}

fn main() {}
