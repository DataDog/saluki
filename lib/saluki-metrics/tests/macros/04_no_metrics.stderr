error: unexpected end of macro invocation
 --> tests/macros/04_no_metrics.rs:5:24
  |
5 |     prefix => basic_usage,
  |                           ^ missing tokens in macro arguments
  |
note: while trying to match `metrics`
 --> src/macros.rs
  |
  |     (name => $name:ident, prefix => $prefix:ident, metrics => [$($metric_type:ident($metric_name:ident)),+ $(,)?] $(,)?) => {
  |                                                    ^^^^^^^
